lib/cusparse/interfaces.jl

```diff
@@ -127,6 +127,73 @@ function Base.:(-)(A::CuSparseMatrix, B::CuSparseMatrixCSR)
     return geam(one(eltype(A)), csrA, -one(eltype(A)), B, 'O')
 end
 
+function Base.reshape(A::CuSparseMatrixCOO{T,M}, dims::NTuple{N,Int}) where {T,N,M}
+    nrows, ncols = size(A)
+    flat_indices = nrows .* (A.colInd .- 1) .+ A.rowInd .- 1
+    new_col, new_row = div.(flat_indices, dims[1]) .+ 1, rem.(flat_indices, dims[1]) .+ 1
+    sparse(new_row, new_col, A.nzVal, dims[1], length(dims) == 1 ? 1 : dims[2], fmt = :coo)
+end
+
+function LinearAlgebra.mul!(Y::CuSparseMatrixCSR{T,M}, A::CuSparseMatrixCSR{T,M}, B::CuSparseMatrixCSR{T,M}) where {T,M}
+    CUSPARSE.version() < v"11.5.1" && throw(ErrorException("This operation is not supported by the current CUDA version."))
+    mm!('N', 'N', one(T), A, B, zero(T), Y, 'O')
+end
+function LinearAlgebra.mul!(Y::CuSparseMatrixCOO{T,M}, A::CuSparseMatrixCOO{T,M}, B::CuSparseMatrixCOO{T,M}) where {T,M}
+    Y2 = CuSparseMatrixCSR(spzeros(T, size(A, 1), size(B, 2)))
+    A2 = CuSparseMatrixCSR(A)
+    B2 = CuSparseMatrixCSR(B)
+    mul!(Y2, A2, B2)
+    copyto!(Y, CuSparseMatrixCOO(Y2))
+end
+function LinearAlgebra.mul!(Y::CuSparseMatrixCSC{T,M}, A::CuSparseMatrixCSC{T,M}, B::CuSparseMatrixCSC{T,M}) where {T,M}
+    Y2 = CuSparseMatrixCSR(spzeros(T, size(A, 1), size(B, 2)))
+    A2 = CuSparseMatrixCSR(A)
+    B2 = CuSparseMatrixCSR(B)
+    mul!(Y2, A2, B2)
+    copyto!(Y, CuSparseMatrixCSC(Y2))
+end
+function LinearAlgebra.:(*)(A::CuSparseMatrixCOO{T,M}, B::CuSparseMatrixCOO{T,M}) where {T,M}
+    Y = CuSparseMatrixCOO(spzeros(T, size(A, 1), size(B, 2)))
+    mul!(Y, A, B)
+end
+
+function SparseArrays.droptol!(A::CuSparseMatrixCOO{T,M}, tol::Real) where {T,M}
+    mask = abs.(A.nzVal) .> tol
+    rows = A.rowInd[mask]
+    cols = A.colInd[mask]
+    datas = A.nzVal[mask]
+    B = sparse(rows, cols, datas, size(A, 1), size(A, 2), fmt = :coo)
+    copyto!(A, B)
+end
+
+for SparseMatrixType in [:CuSparseMatrixCSC, :CuSparseMatrixCSR]
+    @eval begin
+        Base.reshape(A::$SparseMatrixType{T,M}, dims::NTuple{N,Int}) where {T,N,M} = $SparseMatrixType( reshape(CuSparseMatrixCOO(A), dims) )
+
+        function LinearAlgebra.:(*)(A::$SparseMatrixType{T,M}, B::$SparseMatrixType{T,M}) where {T,M}
+            Y = $SparseMatrixType(spzeros(T, size(A, 1), size(B, 2)))
+            mul!(Y, A, B)
+        end
+
+        function LinearAlgebra.adjoint(A::$SparseMatrixType{T,M}) where {T,M}
+            cscA = CuSparseMatrixCSC(conj(A))
+            $SparseMatrixType( CuSparseMatrixCSR(cscA.colPtr, cscA.rowVal, cscA.nzVal, reverse(size(cscA))) )
+        end
+
+        function LinearAlgebra.transpose(A::$SparseMatrixType{T,M}) where {T,M}
+            cscA = CuSparseMatrixCSC(A)
+            $SparseMatrixType( CuSparseMatrixCSR(cscA.colPtr, cscA.rowVal, cscA.nzVal, reverse(size(cscA))) )
+        end
```

Comment:
It will overwrite the lazy `transpose` and `adjoint`, no?

Reply:
Until now, there is not transpose or adjoint functions. If you try to do `adjoint` to any CuSparseMatrix you will get an error.

Reply:
![Capture d’écran du 2022-10-07 01-12-08](https://user-images.githubusercontent.com/35051714/194472492-903d9d73-31bf-47c1-a43f-babba03b3b70.png)

Reply:
I get this error instead

```julia
adjoint(A2)
5×5 adjoint(::CuSparseMatrixCSR{Float64, Int32}) with eltype Float64:
Error showing value of type Adjoint{Float64, CuSparseMatrixCSR{Float64, Int32}}:
ERROR: Scalar indexing is disallowed.
Invocation of getindex resulted in scalar indexing of a GPU array.
This is typically caused by calling an iterating implementation of a method.
Such implementations *do not* execute on the GPU, but very slowly on the CPU,
and therefore are only permitted from the REPL for prototyping purposes.
If you did intend to index this array, annotate the caller with @allowscalar.
Stacktrace:
  [1] error(s::String)
    @ Base ./error.jl:35
  [2] assertscalar(op::String)
    @ GPUArraysCore ~/.julia/packages/GPUArraysCore/lojQM/src/GPUArraysCore.jl:87
  [3] getindex(xs::CuArray{Int32, 1, CUDA.Mem.DeviceBuffer}, I::Int64)
    @ GPUArrays ~/.julia/packages/GPUArrays/fqD8z/src/host/indexing.jl:9
  [4] getindex(A::CUDA.CUSPARSE.CuSparseMatrixCSR{Float64, Int32}, i0::Int64, i1::Int64)
    @ CUDA.CUSPARSE ~/.julia/dev/CUDA/lib/cusparse/array.jl:321
  [5] getindex
    @ ~/julia-1.8.1/share/julia/stdlib/v1.8/LinearAlgebra/src/adjtrans.jl:180 [inlined]
  [6] isassigned(::Adjoint{Float64, CUDA.CUSPARSE.CuSparseMatrixCSR{Float64, Int32}}, ::Int64, ::Int64)
    @ Base ./abstractarray.jl:565
  [7] alignment(io::IOContext{Base.TTY}, X::AbstractVecOrMat, rows::Vector{Int64}, cols::Vector{Int64}, cols_if_complete::Int64, cols_otherwise::Int64, sep::Int64, ncols::Int64)
    @ Base ./arrayshow.jl:68
  [8] _print_matrix(io::IOContext{Base.TTY}, X::AbstractVecOrMat, pre::String, sep::String, post::String, hdots::String, vdots::String, ddots::String, hmod::Int64, vmod::Int64, rowsA::UnitRange{Int64}, colsA::UnitRange{Int64})
    @ Base ./arrayshow.jl:207
  [9] print_matrix(io::IOContext{Base.TTY}, X::Adjoint{Float64, CUDA.CUSPARSE.CuSparseMatrixCSR{Float64, Int32}}, pre::String, sep::String, post::String, hdots::String, vdots::String, ddots::String, hmod::Int64, vmod::Int64) (repeats 2 times)
    @ Base ./arrayshow.jl:171
 [10] print_array
    @ ./arrayshow.jl:358 [inlined]
 [11] show(io::IOContext{Base.TTY}, #unused#::MIME{Symbol("text/plain")}, X::Adjoint{Float64, CUDA.CUSPARSE.CuSparseMatrixCSR{Float64, Int32}})
    @ Base ./arrayshow.jl:399
 [12] (::REPL.var"#43#44"{REPL.REPLDisplay{REPL.LineEditREPL}, MIME{Symbol("text/plain")}, Base.RefValue{Any}})(io::Any)
    @ REPL ~/julia-1.8.1/share/julia/stdlib/v1.8/REPL/src/REPL.jl:267
 [13] with_repl_linfo(f::Any, repl::REPL.LineEditREPL)
    @ REPL ~/julia-1.8.1/share/julia/stdlib/v1.8/REPL/src/REPL.jl:521
 [14] display(d::REPL.REPLDisplay, mime::MIME{Symbol("text/plain")}, x::Any)
    @ REPL ~/julia-1.8.1/share/julia/stdlib/v1.8/REPL/src/REPL.jl:260
 [15] display(d::REPL.REPLDisplay, x::Any)
    @ REPL ~/julia-1.8.1/share/julia/stdlib/v1.8/REPL/src/REPL.jl:272
 [16] display(x::Any)
    @ Base.Multimedia ./multimedia.jl:328
 [17] #invokelatest#2
    @ ./essentials.jl:729 [inlined]
 [18] invokelatest
    @ ./essentials.jl:726 [inlined]
 [19] print_response(errio::IO, response::Any, show_value::Bool, have_color::Bool, specialdisplay::Union{Nothing, AbstractDisplay})
    @ REPL ~/julia-1.8.1/share/julia/stdlib/v1.8/REPL/src/REPL.jl:296
 [20] (::REPL.var"#45#46"{REPL.LineEditREPL, Pair{Any, Bool}, Bool, Bool})(io::Any)
    @ REPL ~/julia-1.8.1/share/julia/stdlib/v1.8/REPL/src/REPL.jl:278
 [21] with_repl_linfo(f::Any, repl::REPL.LineEditREPL)
    @ REPL ~/julia-1.8.1/share/julia/stdlib/v1.8/REPL/src/REPL.jl:521
 [22] print_response(repl::REPL.AbstractREPL, response::Any, show_value::Bool, have_color::Bool)
    @ REPL ~/julia-1.8.1/share/julia/stdlib/v1.8/REPL/src/REPL.jl:276
 [23] (::REPL.var"#do_respond#66"{Bool, Bool, REPL.var"#77#87"{REPL.LineEditREPL, REPL.REPLHistoryProvider}, REPL.LineEditREPL, REPL.LineEdit.Prompt})(s::REPL.LineEdit.MIState, buf::Any, ok::Bool)
    @ REPL ~/julia-1.8.1/share/julia/stdlib/v1.8/REPL/src/REPL.jl:857
 [24] #invokelatest#2
    @ ./essentials.jl:729 [inlined]
 [25] invokelatest
    @ ./essentials.jl:726 [inlined]
 [26] run_interface(terminal::REPL.Terminals.TextTerminal, m::REPL.LineEdit.ModalInterface, s::REPL.LineEdit.MIState)
    @ REPL.LineEdit ~/julia-1.8.1/share/julia/stdlib/v1.8/REPL/src/LineEdit.jl:2510
 [27] run_frontend(repl::REPL.LineEditREPL, backend::REPL.REPLBackendRef)
    @ REPL ~/julia-1.8.1/share/julia/stdlib/v1.8/REPL/src/REPL.jl:1248
 [28] (::REPL.var"#49#54"{REPL.LineEditREPL, REPL.REPLBackendRef})()
    @ REPL ./task.jl:484
```

I think it is related to the showing call.

Reply:
The issue is with `show`. It's quite hard to implement it without scalar indexing.
`adjoint` and `transpose` are already implemented, we are just unable to show the output in the REPL.
```julia
adjoint_A = adjoint(A2);
```

Reply:
Ok, and why you can see both the transpose and adjoint?
Anyway, I will remove both transpose and adjoint functions, and I will implement the mul! function also for these cases.

Reply:
Probably because a global `CUDA.allowscalar(true)`? Maybe we should default to this in REPL tasks...

Reply:
Yes this was the cause.

I removed both adjoint and transpose functions, in order to reuse the lazy versions. However I noticed that I can't convert the type of a transpose/adjoint matrix when it is rectangular

```julia
a = sprand(10, 5, 0.5)
A = CuSparseMatrixCSC(a)
prova = transpose(A)
CuSparseMatrixCSR(prova)


CUSPARSEError: kernel launch failure (code 6, CUSPARSE_STATUS_EXECUTION_FAILED)

Stacktrace:
  [1] throw_api_error(res::CUDA.CUSPARSE.cusparseStatus_t)
    @ CUDA.CUSPARSE ~/.julia/dev/CUDA/lib/cusparse/error.jl:24
  [2] macro expansion
    @ ~/.julia/dev/CUDA/lib/cusparse/error.jl:37 [inlined]
  [3] cusparseCsr2cscEx2(handle::Ptr{Nothing}, m::Int64, n::Int64, nnz::Int32, csrVal::CuArray{Float64, 1, CUDA.Mem.DeviceBuffer}, csrRowPtr::CuArray{Int32, 1, CUDA.Mem.DeviceBuffer}, csrColInd::CuArray{Int32, 1, CUDA.Mem.DeviceBuffer}, cscVal::CuArray{Float64, 1, CUDA.Mem.DeviceBuffer}, cscColPtr::CuArray{Int32, 1, CUDA.Mem.DeviceBuffer}, cscRowInd::CuArray{Int32, 1, CUDA.Mem.DeviceBuffer}, valType::Type, copyValues::CUDA.CUSPARSE.cusparseAction_t, idxBase::Char, alg::CUDA.CUSPARSE.cusparseCsr2CscAlg_t, buffer::CuArray{UInt8, 1, CUDA.Mem.DeviceBuffer})
    @ CUDA.CUSPARSE ~/.julia/dev/CUDA/lib/utils/call.jl:26
  [4] (::CUDA.CUSPARSE.var"#2141#2145"{Char, CuSparseMatrixCSR{Float64, Int32}, CuArray{Float64, 1, CUDA.Mem.DeviceBuffer}, CuArray{Int32, 1, CUDA.Mem.DeviceBuffer}, CuArray{Int32, 1, CUDA.Mem.DeviceBuffer}, Int64, Int64})(buffer::CuArray{UInt8, 1, CUDA.Mem.DeviceBuffer})
    @ CUDA.CUSPARSE ~/.julia/dev/CUDA/lib/cusparse/conversions.jl:154
  [5] with_workspace(f::CUDA.CUSPARSE.var"#2141#2145"{Char, CuSparseMatrixCSR{Float64, Int32}, CuArray{Float64, 1, CUDA.Mem.DeviceBuffer}, CuArray{Int32, 1, CUDA.Mem.DeviceBuffer}, CuArray{Int32, 1, CUDA.Mem.DeviceBuffer}, Int64, Int64}, eltyp::Type{UInt8}, size::CUDA.CUSPARSE.var"#bufferSize#2144"{Char, CuSparseMatrixCSR{Float64, Int32}, CuArray{Float64, 1, CUDA.Mem.DeviceBuffer}, CuArray{Int32, 1, CUDA.Mem.DeviceBuffer}, CuArray{Int32, 1, CUDA.Mem.DeviceBuffer}, Int64, Int64}, fallback::Nothing; keep::Bool)
    @ CUDA.APIUtils ~/.julia/dev/CUDA/lib/utils/call.jl:77
  [6] with_workspace
    @ ~/.julia/dev/CUDA/lib/utils/call.jl:56 [inlined]
  [7] #with_workspace#1
    @ ~/.julia/dev/CUDA/lib/utils/call.jl:53 [inlined]
  [8] with_workspace (repeats 2 times)
    @ ~/.julia/dev/CUDA/lib/utils/call.jl:53 [inlined]
  [9] #_#2140
    @ ~/.julia/dev/CUDA/lib/cusparse/conversions.jl:153 [inlined]
 [10] CuSparseMatrixCSC
    @ ~/.julia/dev/CUDA/lib/cusparse/conversions.jl:139 [inlined]
 [11] CuSparseMatrixCSC
    @ ~/.julia/dev/CUDA/lib/cusparse/array.jl:381 [inlined]
```

But I think I need this implementation. The only working way I found is by writing `transpose(CuSparseMatrixCSR(parent(prova)))` but, obviously, it is not convenient.